Make Split Mongo's CachedDescriptorSystem load field-data service on demand

bradenmacdonald opened the pull request:
## Description

This is a comprehensive fix for https://github.com/openedx/modular-learning/issues/56

What this does:

Currently, whenever any XBlock is loaded, the `FieldData` instance for that XBlock is loaded by `CachingDescriptorSystem` and injected into `block._field_data` via the XBlock's constructor, although [doing so is deprecated](https://github.com/openedx/XBlock/blob/e89cbc5c670d118d7a9e8ff0ef302eba10b815ec/xblock/mixins.py#L200-L211).

With this PR, I've changed split mongo's `CachingDescriptorSystem` to work more like [the other new XBlock runtime that I made](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/xblock/runtime/runtime.py): The `FieldData` API is only available via the `field-data` service, and is only loaded on demand whenever some field data is first accessed. Then it is cached so there is no overhead for subsequent field reads.

Note: Old/draft Mongo also has its own implementation of `CachingDescriptorSystem`, and I modified it as little as possible, so it still mostly works as it did before.

~~In some cases where field data was never actually being read, this PR has slightly reduced the number of mongo calls being made for certain parts of the test suite.~~ (Edit: actually this benefit came from using `block.scope_ids` as a key for field data, but that approach turned out to be infeasible due to a number of crazy ways that the current system changes `scope_ids` throughout the lifetime of a block - during such diverse activities as bind_for_student, CCX, imports, and saving changes - and several things depend on that happening.)

## Supporting information

See linked bug report.

## Testing instructions

See linked bug report.

## Deadline

None

## Other information

Private-ref: MNG-3711

bradenmacdonald committed:
fix: field data should be loaded on-demand, not pushed into block init

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald!

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.



<!-- comment:welcome-core-committer -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bradenmacdonald committed:
temp: continued

bradenmacdonald committed:
temp: update test case, uses one less query since field-data is on demand

bradenmacdonald committed:
temp: clarify why we don't pass field_data to constructor

bradenmacdonald committed:
temp: continue

bradenmacdonald committed:
temp: update query counts

bradenmacdonald committed:
fix: failing test, not really sure why this worked before without modulestore.update_item()

bradenmacdonald committed:
temp: update query counts

bradenmacdonald commented:
@ormsbee @Agrendalath What do you think of this approach to solving https://github.com/openedx/modular-learning/issues/56? Too risky?

The diff looks significant but it's _mostly_ just moving code from the `xblock_from_json` method into a new `_init_field_data_for_block` method which gets deferred until someone tries to read from an XBlock field.

This does solve the bug from https://github.com/openedx/modular-learning/issues/56

ormsbee commented:
I took a quick look and it seems pretty reasonable. Though changing something this fundamental would warrant special attention from @jristau1984 and TNL folks during rollout.

As a side note, I wonder if this level of caching would fix some of CCX's excessive mongo queries.

Agrendalath commented:
@bradenmacdonald, I'll take a more careful look at this tomorrow, but it seems reasonable and works well from initial checks.

Agrendalath commented on xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -167,12 +170,83 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
 
-        # If no usage id is provided, generate an in-memory id
-        if block_key is None:
-            block_key = BlockKey(block_data.block_type, LocalId())
```
According to [this call](https://github.com/openedx/edx-platform/blob/8ee1f66ffbdfe23b58cccc4fe49b01a10f5a0b77/xmodule/modulestore/split_mongo/split.py#L2112), `block_key` can be `None`. I don't know if this happens often in real use-cases, but the [cms/djangoapps/contentstore/tests/test_crud.py::TemplateTests::test_temporary_xblocks](https://github.com/openedx/edx-platform/blob/3a1011bed87c31c6c3e6cb314e35bbae8207cd14/cms/djangoapps/contentstore/tests/test_crud.py#L92-L95) test is failing because of this. I pushed a small fix to Mongo call tests to unblock other pipelines.

We may want to restore the deleted code and move it a few lines up.

bradenmacdonald replied:
Thanks @Agrendalath! I am going to continue work on this today and hope to have all the tests passing soon. I appreciate the help and the feedback!

Agrendalath commented on xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -167,12 +170,83 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
-        # If no usage id is provided, generate an in-memory id
-        if block_key is None:
-            block_key = BlockKey(block_data.block_type, LocalId())
+        definition_id = block_data.definition or LocalId()
```
```suggestion
        # If no definition id is provided, generate an in-memory id
        definition_id = block_data.definition or LocalId()
```

We can move this comment here from `_init_field_data_for_block`.

Agrendalath commented on xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -192,12 +266,6 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
         if definition_id is None:
             definition_id = LocalId()
```
At this point, `definition_id` is no longer used.

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -653,12 +651,13 @@ def bind_for_student(self, user_id, wrappers=None):
 
         if wrappers is None:
             wrappers = []
```
Nit: we can remove this part because we do the `if wrappers` check below.

Side note: I'm wondering if there are any use cases for not specifying the wrappers at all (outside of tests). I haven't found any.

Agrendalath committed:
test: fix the number of Mongo calls

bradenmacdonald committed:
fix: cleanups and bug fix from review

bradenmacdonald committed:
fix: remove the need to patch the KVS during library import

bradenmacdonald committed:
fix: Apparently block.scope_ids gets mutated a lot and isn't a good key

-> bind_for_student() and friends change the user_id
-> DraftVersioningModuleStore.update_item() mutates .location which mutates scope_ids.def_id and scope_ids.usage_id in a way that seems totally wrong

bradenmacdonald committed:
fix: make the initialization more like how it was in the past

bradenmacdonald committed:
fix: some tests still use old mongo and weren't passing

bradenmacdonald committed:
refactor: we don't need _unwrapped_field_data any more

bradenmacdonald commented:
@ormsbee @Agrendalath This is ready for a final review now. I had to make the changes a bit more conservative in some places, so it no longer has any effect on query counts. But hopefully that also means it's less risky.

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -367,7 +379,7 @@ def location(self):
         assert isinstance(value, UsageKey)
         self.scope_ids = self.scope_ids._replace(
-            def_id=value,
+            def_id=value,  # FIXME: this seems wrong... a UsageKey is not a definition key.
```
I see that we're using the `UsageKey` in multiple places:
1. https://github.com/openedx/edx-platform/blob/dac1c5abe1193ce51297d6aa6db32742b2f6cedc/xmodule/modulestore/mongo/base.py#L225-L277
2. https://github.com/openedx/edx-platform/blob/dac1c5abe1193ce51297d6aa6db32742b2f6cedc/xmodule/modulestore/mongo/base.py#L1264-L1271
3. https://github.com/openedx/edx-platform/blob/f13263df4f0e9aff48fbbb7ca04583619317b947/xmodule/modulestore/xml.py#L687 This one is retrieved from: https://github.com/openedx/edx-platform/blob/f13263df4f0e9aff48fbbb7ca04583619317b947/xmodule/modulestore/xml.py#L627
4. https://github.com/openedx/edx-platform/blob/537cfe4e0f822c5dc7ebbd244ae289a33d9a912a/xmodule/modulestore/xml_importer.py#L112-L116

bradenmacdonald replied:
Ah, that explains it, thank you - I see now that old mongo and the XML importer system may use the same key for both, but **split** definitely doesn't. In fact, I think that's what the name "split" refers to, that the definitions are split from the usage. Or maybe it's that the structure is split from the definitions, but it has a similar effect.

Agrendalath commented on xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -259,18 +211,110 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
+                except:
+                    # Don't try again pointlessly every time another field is accessed
+                    self.block_field_datas[block] = None
+                    raise
```
Are we catching this exception somewhere?

bradenmacdonald replied:
I seem to recall occasionally seeing cases where this type of exception is caught and without the line above to make it happen only once, it generates tons of exceptions in the logs... but in general I'm not sure; we shouldn't be catching it, as it's a major error that should cause the request to end.

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: did regression testing with various XBlocks
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

bradenmacdonald committed:
docs: clarify comment that I had added earlier

bradenmacdonald committed:
chore: Update with latest master

bradenmacdonald commented:
@ormsbee could you please give this a final look? And @jristau1984 would you then be able to squash+merge it at a time that's convenient for monitoring the roll out?

ormsbee commented on xmodule/modulestore/xml.py:
```diff
@@ -239,6 +239,24 @@ def load_item(usage_key, for_parent=None):
+        defined for this application.
+        """
+        if field_data:
+            # Currently, *some* XBlocks (those with XmlMixin) use XmlMixin.parse_xml() which instantiates
+            # its own key value store for field data. That is something which should be left to the runtime, for
+            # consistent behavior across all XBlocks, not controlled by individual XBlock implementations.
+            # We cannot just ignore field_data here though, because parse_xml may have pre-loaded data into it that we
+            # would otherwise lose.
```
This is a great warning, but now I'm curious to understand–why does `XmlMixin.parse_xml()` create its own `KvsFieldData`?

bradenmacdonald replied:
I honestly don't know. Possibly because at the time there was no API for asking the runtime to create a new block without passing in a field data instance? (But now there is...)

And yes, I hope this warning will help others; I'm adding it because it was difficult to debug some inconsistent behavior I was seeing until I finally figured out that during import or parsing OLX, some XBlocks were using totally different FieldData than what the runtime was providing for them to use.

ormsbee commented on xmodule/modulestore/xml.py:
```diff
@@ -899,26 +917,10 @@ def get_id(org, library, url_name):  # lint-amnesty, pylint: disable=arguments-d
-        # the next time it will be called
-        lazy.invalidate(library_block, '_unwrapped_field_data')
-        # pylint: disable=protected-access
-        library_block._field_data = inheriting_field_data(InheritanceKeyValueStore(init_dict))
```
I know we do some weird things around [library defaults vs. inheritance](https://github.com/openedx/edx-platform/blob/aa68ea116208c430fce2e1aac14ecc50d476da02/xmodule/modulestore/inheritance.py#L383-L386) for library blocks. Is this part definitely safe to remove?

bradenmacdonald replied:
This code was a hack to monkey patch the FieldData instance on the root library XBlock to be an `InheritanceKeyValueStore` during import, because the author noticed some errors if that wasn't done. But I found that by making the changes I made to [xmodule/modulestore/inheritance.py](https://github.com/openedx/edx-platform/pull/32273/files/aa68ea116208c430fce2e1aac14ecc50d476da02#diff-174364edda579bd95854e17d32c0c2aa59df72c9474c66ea462762c7bc00b587), this hack is unnecessary. All of the related tests in [xmodule/tests/test_import.py](https://github.com/openedx/edx-platform/pull/32273/files/aa68ea116208c430fce2e1aac14ecc50d476da02#diff-761df6793ec276652eedec1d4b3e04bb622d440f2908a2d100c1d910249f1a3b) which are explicitly testing library inheritance are still passing, and now they don't need the monkey-patching anymore.

In any case, there is no mechanism in Studio to set any attributes on the root `library` XBlock, so I believe that even if this changed behavior, it would only affect users who edited the root XBlock OLX to set some inherited fields into the library, which I don't think is something we encourage or support. In fact, I think the link that you included in your comment shows that field data never comes from the root library block anyways - it either comes from the library component or overrides set onto the component once it's used in a course. 

P.S. I believe the quote "as XBlocks don't have this KVS" is referring to how `parse_xml` gives XModules a different KVS than XBlocks during import, causing different behavior at the time this was written for the `course` XModule as for the `library` XBlock.

ormsbee reviewed (commented):
I just have a couple of questions, including one around library content, and if we should be concerned about regressions when importing a course with library content in it.

bradenmacdonald commented:
@jristau1984 would you mind merging at your convenience, when you can keep an eye on things after it rolls out? I don't expect any problems, but it is a bit riskier than our other PRs in that it changes a part of the modulestore xblock runtime.

openedx-webhooks commented:
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/32273&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2023-05-19+20:57&entry.1917517419=2023-06-13+12:05&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
